com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java

```diff
@@ -197,55 +211,66 @@ private void countComponentInResource(Project project, IResource res, Map<String
             addOneComponent(type, target);
             if (isFactoryType(type)) {
                 IResource resource = project.getResource(comp);
-                countComponentsInFactory(project, resource, type);
+                if (!countComponentsInFactory(project, resource, type))
+                    return false;
             }
         }
+        return true;
     }
 
-    private void countComponentsInCollectionRecursively(Project project, CollectionDesc.Builder builder, Map<String, Integer> target) throws IOException, CompileExceptionError {
+    private boolean countComponentsInCollectionRecursively(Project project, CollectionDesc.Builder builder, Map<String, Integer> target) throws IOException, CompileExceptionError {
         for (InstanceDesc inst : builder.getInstancesList()) {
             IResource res = project.getResource(inst.getPrototype());
-            countComponentInResource(project, res, target);
+            if (!countComponentInResource(project, res, target))
+                return false;
         }
         for (EmbeddedInstanceDesc desc : builder.getEmbeddedInstancesList()) {
             byte[] data = desc.getData().getBytes();
             long hash = MurmurHash.hash64(data, data.length);
 
             IResource res = project.getGeneratedResource(hash, "go");
-            countComponentInResource(project, res, target);
+            if (!countComponentInResource(project, res, target))
+                return false;
         }
         for (CollectionInstanceDesc collInst : builder.getCollectionInstancesList()) {
             IResource collResource = project.getResource(collInst.getCollection());
             CollectionDesc.Builder subCollBuilder = CollectionDesc.newBuilder();
             ProtoUtil.merge(collResource, subCollBuilder);
-            countComponentsInCollectionRecursively(project, subCollBuilder, target);
+            if (!countComponentsInCollectionRecursively(project, subCollBuilder, target))
+                return false;
         }
+        return true;
     }
 
     private String replaceComponentName(Project project, String in) {
         return project.replaceExt("." + in).substring(1);
     }
 
     private void countAllComponentTypes(Project project, CollectionDesc.Builder builder) throws IOException, CompileExceptionError {
-        countComponentsInCollectionRecursively(this.project, builder, components);
-
+        boolean isStatic = countComponentsInCollectionRecursively(this.project, builder, components);
```

Comment:
If we can determine if there are no dynamic prototypes, then we can use the component counts.